File: src/bmsx/machine/devices/audio/active_slots.ts
import { IO_APU_ACTIVE_MASK } from '../../bus/io';
import type { Memory } from '../../memory/memory';
import type { ApuEventLatch } from './event_latch';
import type { ApuOutputMixer } from './output';
import type { ApuSelectedSlotLatch } from './selected_slot_latch';
import type { ApuSourceDma } from './source';
import type { ApuSlotAdvanceResult, ApuSlotBank } from './slot_bank';
import {
	APU_EVENT_SLOT_ENDED,
	APU_SLOT_COUNT,
	type ApuAudioSlot,
	type ApuParameterRegisterWords,
	type ApuSlotPhase,
	type ApuVoiceId,
} from './contracts';

export class ApuActiveSlots {
	private readonly advanceResult: ApuSlotAdvanceResult = { ended: false, voiceId: 0, sourceAddr: 0 };
	public constructor(
		private readonly memory: Memory,
		private readonly audioOutput: ApuOutputMixer,
		private readonly sourceDma: ApuSourceDma,
		private readonly eventLatch: ApuEventLatch,
		private readonly slots: ApuSlotBank,
		private readonly selectedSlotLatch: ApuSelectedSlotLatch,
	) {}

	public writeActiveMask(): void {
		this.memory.writeIoValue(IO_APU_ACTIVE_MASK, this.slots.activeMask);
		this.selectedSlotLatch.refresh();
	}

	public setActive(slot: ApuAudioSlot, registerWords: ApuParameterRegisterWords, voiceId: ApuVoiceId): void {
		this.slots.setActive(slot, registerWords, voiceId);
		this.writeActiveMask();
	}

	public stop(slot: ApuAudioSlot): void {
		this.slots.clearSlot(slot);
		this.sourceDma.clearSlot(slot);
		this.writeActiveMask();
	}

	public setPhase(slot: ApuAudioSlot, phase: ApuSlotPhase): void {
		this.slots.setPhase(slot, phase);
		this.writeActiveMask();
	}

	public advance(samples: number): void {
		const result = this.advanceResult;
		for (let slot = 0; slot < APU_SLOT_COUNT; slot += 1) {
			this.slots.advanceSlot(slot, samples, result);
			if (result.ended) {
				this.audioOutput.stopSlot(slot);
				this.emitSlotEvent(slot, result.voiceId, result.sourceAddr);
			}
		}
	}

	private emitSlotEvent(slot: ApuAudioSlot, voiceId: ApuVoiceId, sourceAddr: number): void {
		if (this.slots.voiceId(slot) !== voiceId) {
			return;
		}
		this.stop(slot);
		this.eventLatch.emit(APU_EVENT_SLOT_ENDED, slot, sourceAddr);
	}

}

File: src/bmsx_cpp/machine/devices/audio/active_slots.cpp
#include "machine/devices/audio/active_slots.h"

#include "machine/bus/io.h"
#include "machine/cpu/cpu.h"
#include "machine/devices/audio/event_latch.h"
#include "machine/devices/audio/output.h"
#include "machine/devices/audio/selected_slot_latch.h"
#include "machine/devices/audio/source.h"
#include "machine/devices/audio/slot_bank.h"
#include "machine/memory/memory.h"

namespace bmsx {

ApuActiveSlots::ApuActiveSlots(Memory& memory,
	ApuOutputMixer& audioOutput,
	ApuSourceDma& sourceDma,
	ApuEventLatch& eventLatch,
	ApuSlotBank& slots,
	ApuSelectedSlotLatch& selectedSlotLatch)
	: m_memory(memory)
	, m_audioOutput(audioOutput)
	, m_sourceDma(sourceDma)
	, m_eventLatch(eventLatch)
	, m_slots(slots)
	, m_selectedSlotLatch(selectedSlotLatch) {}

void ApuActiveSlots::writeActiveMask() {
	m_memory.writeIoValue(IO_APU_ACTIVE_MASK, valueNumber(static_cast<double>(m_slots.activeMask())));
	m_selectedSlotLatch.refresh();
}

void ApuActiveSlots::setActive(ApuAudioSlot slot, const ApuParameterRegisterWords& registerWords, ApuVoiceId voiceId) {
	m_slots.setActive(slot, registerWords, voiceId);
	writeActiveMask();
}

void ApuActiveSlots::stop(ApuAudioSlot slot) {
	m_slots.clearSlot(slot);
	m_sourceDma.clearSlot(slot);
	writeActiveMask();
}

void ApuActiveSlots::setPhase(ApuAudioSlot slot, ApuSlotPhase phase) {
	m_slots.setPhase(slot, phase);
	writeActiveMask();
}

void ApuActiveSlots::advance(i64 samples) {
	for (ApuAudioSlot slot = 0; slot < APU_SLOT_COUNT; slot += 1u) {
		const ApuSlotAdvanceResult result = m_slots.advanceSlot(slot, samples);
		if (result.ended) {
			m_audioOutput.stopSlot(slot);
			emitSlotEvent(slot, result.voiceId, result.sourceAddr);
		}
	}
}

void ApuActiveSlots::emitSlotEvent(ApuAudioSlot slot, ApuVoiceId voiceId, u32 sourceAddr) {
	if (m_slots.voiceId(slot) != voiceId) {
		return;
	}
	stop(slot);
	m_eventLatch.emit(APU_EVENT_SLOT_ENDED, slot, sourceAddr);
}

} // namespace bmsx
